x-pack/plugins/cases/server/connectors/cases/retry_service.ts

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { CasesConnectorError } from './cases_connector_error';
+import type { BackoffStrategy, BackoffFactory } from './types';
+
+export class CaseConnectorRetryService {
+  private maxAttempts: number = 10;
+  /**
+   * 409 - Conflict
+   * 429 - Too Many Requests
+   * 503 - ES Unavailable
+   *
+   * Full list of errors: packages/core/saved-objects/core-saved-objects-server/src/saved_objects_error_helpers.ts
+   */
+  private readonly RETRY_ERROR_STATUS_CODES: number[] = [409, 429, 503];
+  private readonly backOffStrategy: BackoffStrategy;
+
+  private timer: NodeJS.Timeout | null = null;
+  private attempt: number = 0;
+
+  constructor(backOffFactory: BackoffFactory, maxAttempts: number = 10) {
+    this.backOffStrategy = backOffFactory.create();
+    this.maxAttempts = maxAttempts;
+  }
+
+  public async retryWithBackoff<T>(cb: () => Promise<T>): Promise<T> {
+    try {
+      const res = await cb();
+      return res;
+    } catch (error) {
+      if (this.shouldRetry() && this.isRetryableError(error)) {
+        this.stop();
+        this.attempt++;
+
+        await this.delay();
+        return this.retryWithBackoff(cb);
+      }
+
+      throw error;
+    }
+  }
+
+  private shouldRetry() {
+    return this.attempt < this.maxAttempts;
+  }
+
+  private isRetryableError(error: Error) {
+    if (
+      error instanceof CasesConnectorError &&
+      this.RETRY_ERROR_STATUS_CODES.includes(error.statusCode)
+    ) {
+      return true;
+    }
+
+    return false;
+  }
+
+  private async delay() {
+    const ms = this.backOffStrategy.nextBackOff();
+
+    return new Promise((resolve) => {
+      this.timer = setTimeout(resolve, ms);
+    });
+  }
+
+  private stop(): void {
+    if (this.timer !== null) {
```

Comment:
nit: I think this check is unnecessary, `clearTimeout` [will just do nothing](https://developer.mozilla.org/en-US/docs/Web/API/clearTimeout).

Reply:
I did not know about it, thanks!

Reply:
I get this TS error. `clearTimeout ` accepts a number or `undefined`. I will keep it as it is, ok?

<img width="888" alt="Screenshot 2024-01-11 at 4 55 39 PM" src="https://github.com/elastic/kibana/assets/7871006/afe6cf14-5173-4f58-b725-c9659a2357b7">

Reply:
Weird when i tried it locally it didn't show anything. nevermind then!